File: apps/ios/KommsCore/Sources/KommsCore/NativeWakePolicy.swift
import Foundation

/// Recipient-selected static APNs profile.
public enum NativeWakePreference: String, Codable, Sendable {
    case disabled
    case backgroundOnly
    case genericVisible
}

/// Alert authorization relevant to the visible profile.
public enum NativeWakePermission: Equatable, Sendable {
    case notRequired
    case granted
    case denied
}

/// Platform action required after one lifecycle transition.
public enum NativeWakeAction: Equatable, Sendable {
    case none
    case register
    case revoke
}

/// Secret-free facts used by the iOS lifecycle decision table.
public struct NativeWakeSnapshot: Equatable, Sendable {
    public var mode: String
    public var gatewayCount: Int
    public var preference: NativeWakePreference
    public var permission: NativeWakePermission
    public var backgroundRefreshAvailable: Bool
    public var tokenDigest: String?
    public var advertised: Bool

    public init(
        mode: String,
        gatewayCount: Int,
        preference: NativeWakePreference,
        permission: NativeWakePermission,
        backgroundRefreshAvailable: Bool,
        tokenDigest: String?,
        advertised: Bool
    ) {
        self.mode = mode
        self.gatewayCount = gatewayCount
        self.preference = preference
        self.permission = permission
        self.backgroundRefreshAvailable = backgroundRefreshAvailable
        self.tokenDigest = tokenDigest
        self.advertised = advertised
    }
}

/// One bounded platform decision.
public struct NativeWakeDecision: Equatable, Sendable {
    public var action: NativeWakeAction
    public var advertise: Bool

    public init(action: NativeWakeAction, advertise: Bool) {
        self.action = action
        self.advertise = advertise
    }
}

/// Pure lifecycle rules shared by the iOS app and host tests.
public enum NativeWakePolicy {
    public static func decide(
        previous: NativeWakeSnapshot?,
        current: NativeWakeSnapshot,
        forceRefresh: Bool = false
    ) -> NativeWakeDecision {
        precondition(["standard", "private", "sovereign"].contains(current.mode))
        let permissionAllows =
            current.preference != .genericVisible || current.permission == .granted
        let executionAllows =
            current.preference != .backgroundOnly || current.backgroundRefreshAvailable
        let eligible =
            current.mode != "sovereign"
            && (1...4).contains(current.gatewayCount)
            && current.preference != .disabled
            && current.tokenDigest != nil
            && permissionAllows
            && executionAllows
        guard eligible else {
            return NativeWakeDecision(
                action: previous?.advertised == true || current.advertised ? .revoke : .none,
                advertise: false)
        }
        let changed =
            previous == nil
            || previous?.tokenDigest != current.tokenDigest
            || previous?.preference != current.preference
            || previous?.mode != current.mode
            || previous?.gatewayCount != current.gatewayCount
            || previous?.advertised != true
        return NativeWakeDecision(
            action: changed || forceRefresh ? .register : .none,
            advertise: true)
    }

    /// Accept only the two static content-free APNs dictionaries.
    public static func acceptsStaticPayload(
        contentAvailable: Int,
        alertTitle: String?,
        alertBody: String?,
        sound: String?,
        applicationKeys: Set<String>
    ) -> Bool {
        guard contentAvailable == 1, applicationKeys.isEmpty else { return false }
        if alertTitle == nil && alertBody == nil && sound == nil { return true }
        return alertTitle == "Komms"
            && alertBody == "New activity"
            && sound == "default"
    }

    /// iOS does not execute background work after force-quit and may suppress
    /// it when Background App Refresh is disabled.
    public static func canRunBackgroundCollection(
        backgroundRefreshAvailable: Bool,
        forceQuit: Bool
    ) -> Bool {
        backgroundRefreshAvailable && !forceQuit
    }
}

File: apps/android/core/src/main/kotlin/komms/core/NativeWakePolicy.kt
package komms.core

/** Recipient-selected static native-notification profile. */
enum class NativeWakePreference {
    DISABLED,
    BACKGROUND_ONLY,
    GENERIC_VISIBLE,
}

/** Android notification authorization relevant to a visible wake. */
enum class NativeWakePermission {
    NOT_REQUIRED,
    GRANTED,
    DENIED,
}

/** Platform action required after a lifecycle transition. */
enum class NativeWakeAction {
    NONE,
    REGISTER,
    REVOKE,
}

/** Secret-free lifecycle facts used to decide native-wake capability state. */
data class NativeWakeSnapshot(
    val playBuild: Boolean,
    val mode: String,
    val gatewayCount: Int,
    val preference: NativeWakePreference,
    val permission: NativeWakePermission,
    val tokenDigest: String?,
    val advertised: Boolean,
)

/** One bounded platform decision. */
data class NativeWakeDecision(
    val action: NativeWakeAction,
    val advertise: Boolean,
    val highPriorityAllowed: Boolean,
)

/**
 * Android lifecycle policy kept independent of Firebase and Android APIs so
 * Google-free builds and host tests consume the same decision table.
 */
object NativeWakePolicy {
    fun decide(
        previous: NativeWakeSnapshot?,
        current: NativeWakeSnapshot,
        forceRefresh: Boolean = false,
    ): NativeWakeDecision {
        require(current.mode in setOf("standard", "private", "sovereign"))
        val eligible =
            current.playBuild &&
                current.mode != "sovereign" &&
                current.gatewayCount in 1..4 &&
                current.preference != NativeWakePreference.DISABLED &&
                current.tokenDigest != null &&
                (
                    current.preference != NativeWakePreference.GENERIC_VISIBLE ||
                        current.permission == NativeWakePermission.GRANTED
                    )
        if (!eligible) {
            return NativeWakeDecision(
                action = if (previous?.advertised == true || current.advertised) {
                    NativeWakeAction.REVOKE
                } else {
                    NativeWakeAction.NONE
                },
                advertise = false,
                highPriorityAllowed = false,
            )
        }
        val changed =
            previous == null ||
                previous.tokenDigest != current.tokenDigest ||
                previous.preference != current.preference ||
                previous.mode != current.mode ||
                previous.gatewayCount != current.gatewayCount ||
                !previous.advertised
        return NativeWakeDecision(
            action = if (changed || forceRefresh) NativeWakeAction.REGISTER else NativeWakeAction.NONE,
            advertise = true,
            highPriorityAllowed = current.preference == NativeWakePreference.GENERIC_VISIBLE,
        )
    }

    /** Accept only the two content-free FCM shapes emitted by the gateway. */
    fun acceptsStaticPayload(
        data: Map<String, String>,
        notificationTitle: String?,
        notificationBody: String?,
    ): Boolean {
        if (data != mapOf("wake" to "1")) return false
        return (notificationTitle == null && notificationBody == null) ||
            (notificationTitle == "Komms" && notificationBody == "New activity")
    }

    /** Continuation is scheduled only for bounded unprocessed core work. */
    fun shouldScheduleContinuation(
        remaining: Boolean,
        sessionUnlocked: Boolean,
        playBuild: Boolean,
    ): Boolean = remaining && sessionUnlocked && playBuild
}
